packages/hardhat-core/src/internal/util/download.ts

```diff
@@ -27,47 +25,48 @@ export async function download(
   timeoutMillis = 10000
 ) {
   const { pipeline } = await import("stream");
-  const { default: fetch } = await import("node-fetch");
+  const { getGlobalDispatcher, ProxyAgent, request } = await import("undici");
   const streamPipeline = util.promisify(pipeline);
-  const fetchOptions: FetchOptions = {
-    timeout: timeoutMillis,
-    agent: undefined,
-  };
 
-  // Check if Proxy is set https
-  if (process.env.HTTPS_PROXY !== undefined) {
-    // Create the proxy from the environment variables
-    const proxy: string = process.env.HTTPS_PROXY;
-    fetchOptions.agent = new HttpsProxyAgent.HttpsProxyAgent(proxy);
-  }
+  function chooseDispatcher() {
+    if (process.env.HTTPS_PROXY !== undefined) {
+      return new ProxyAgent(process.env.HTTPS_PROXY);
+    }
 
-  // Check if Proxy is set http and `fetchOptions.agent` was not already set for https
-  if (
-    process.env.HTTP_PROXY !== undefined &&
-    fetchOptions.agent === undefined
-  ) {
-    // Create the proxy from the environment variables
-    const proxy: string = process.env.HTTP_PROXY;
-    fetchOptions.agent = new HttpsProxyAgent.HttpsProxyAgent(proxy);
+    if (process.env.HTTP_PROXY !== undefined) {
+      return new ProxyAgent(process.env.HTTP_PROXY);
+    }
+
+    return getGlobalDispatcher();
   }
 
   // Fetch the url
-  const response = await fetch(url, fetchOptions);
+  const response = await request(url, {
+    dispatcher: chooseDispatcher(),
+    headersTimeout: timeoutMillis,
+    maxRedirections: 10,
+    method: "GET",
+    headers: {
+      "User-Agent": `hardhat ${hardhatVersion}`,
+    },
+  });
 
-  if (response.ok && response.body !== null) {
+  if (
+    response.statusCode >= 200 &&
+    response.statusCode <= 299 &&
+    response.body !== null
+  ) {
     const tmpFilePath = resolveTempFileName(filePath);
     await fsExtra.ensureDir(path.dirname(filePath));
 
     await streamPipeline(response.body, fs.createWriteStream(tmpFilePath));
     return fsExtra.move(tmpFilePath, filePath, { overwrite: true });
   }
 
-  // Consume the response stream and discard its result
-  // See: https://github.com/node-fetch/node-fetch/issues/83
-  const _discarded = await response.arrayBuffer();
-
   // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
   throw new Error(
-    `Failed to download ${url} - ${response.statusText} received`
+    `Failed to download ${url} - ${
+      response.statusCode
+    } received. ${await response.body.text()}`
```

Comment:
If the `if` was not entered because `response.body` was null, this will throw while building the error message.